### PR TITLE
Morrowinventory

### PR DESCRIPTION
**Pull request by BobJoga**
## About The Pull Request

image

Makes underwear work as actual items, adds slots for them, adds a second ear slot.

Reworks accessory code so you can have multiple accessories.

Adds a new toggle for the new UI slots, so you don't need to have the screen cluttered to access most important clothing and such.

This will probably be part 1 of an entire series of PRs just about me going completely insane with inventory code.

Of course, my ultimate intention is for the full inventory to end up looking like this:
image

**To-do list:**
- [x] Make items for every currently existing underwear type
- [x] Make headsets work in the extra slot
- [x] Update the clothesmate for the new underwear system
- [x] Make loadout datums for every one of these underwear items
- [x] Find a way to create a color selection for underwear in the loadout
- [x] Add a functional wrist slot

## Why It's Good For The Game

I like deep inventory systems.

## Changelog
:cl:
del: Sprite accessory underwear is gone, replaced by proper underwear items.
del: Wardrobe no longer has any functionality until i can find a way to restore it.
add: Added underwear to the loadout
add: Added underwear slots
add: Added wrist slot
add: Added extra ear slot
add: Added toggle for extra inventory slots
add: Clothing vendors now sell underwear
balance: Prayer beads can now be equipped in the wrist slot
refactor: Uniforms can now have multiple accessories
/:cl:

### CODE REVIEW

**Comment by ShadeAware** [comment]
Oh YOOO this actually looks kind of nice

**Comment by TheSpaghetti** [comment]
ah yes, THICC inventory

**Comment by DarkWolfInsanity** [comment]
Diablo Inventory where items can take up more than one slot and have weight when

**Comment by BobJoga** [comment]
> Diablo Inventory where items can take up more than one slot and have weight when

Cit already did the code for the latter but it's kinda fucky wucky. About the first, yes when i get to porting rig suits.

**Comment by DarkWolfInsanity** [comment]
hold the fuck up. wrist slots? We get fuckin watches and bracelets and super secret micro missiles now?

**Comment by KasparoVy** [comment]
Having to spend loadout points on underwear lol, fuuuck.

GOING COMMANDO

**Comment by BobJoga** [comment]
> hold the fuck up. wrist slots? We get fuckin watches and bracelets and super secret micro missiles now?

yes

**Comment by BobJoga** [comment]
> Having to spend loadout points on underwear lol, fuuuck.
> 
> GOING COMMANDO

yes

File: code/game/objects/structures/dresser.dm [hunk 1]
```diff
@@ -18,65 +31,87 @@
 /obj/structure/dresser/deconstruct(disassembled = TRUE)
 	new /obj/item/stack/sheet/mineral/wood(drop_location(), 10)
 	qdel(src)
-
+/*
+//skyrat edit whole proc got modified pretty much
```
Comment (Gandalf2k15) [inline code/game/objects/structures/dresser.dm:35]: Reference where the proc was moved to.

File: code/modules/clothing/suits/_suits.dm [hunk 2]
```diff
@@ -23,9 +23,12 @@
 		var/mob/living/carbon/human/M = loc
 		if(ishuman(M) && M.w_uniform)
 			var/obj/item/clothing/under/U = M.w_uniform
-			if(istype(U) && U.attached_accessory)
-				var/obj/item/clothing/accessory/A = U.attached_accessory
-				if(A.above_suit)
+			if(istype(U) && length(U.attached_accessories))
```
Comment (Gandalf2k15) [inline code/modules/clothing/suits/_suits.dm:26]: skyrat edit

File: code/modules/clothing/under/_under.dm [hunk 3]
```diff
@@ -15,7 +15,8 @@
 	var/adjusted = NORMAL_STYLE
 	var/alt_covers_chest = FALSE // for adjusted/rolled-down jumpsuits, FALSE = exposes chest and arms, TRUE = exposes arms only
 	var/dummy_thick = FALSE // is able to hold accessories on its item
-	var/obj/item/clothing/accessory/attached_accessory
+	var/list/obj/item/clothing/accessory/attached_accessories = list()
```
Comment (Gandalf2k15) [inline code/modules/clothing/under/_under.dm:18]: Skyrat edit documentation missing in this whole file.

File: code/modules/clothing/under/accessories.dm [hunk 4]
```diff
@@ -17,15 +23,10 @@
 			return FALSE
 		U.TakeComponent(storage)
 		detached_pockets = storage
-	U.attached_accessory = src
+	U.attached_accessories |= src
```
Comment (Gandalf2k15) [inline code/modules/clothing/under/accessories.dm:26]: SKYRAT EDIT

File: code/modules/clothing/under/accessories.dm [hunk 5]
```diff
@@ -45,19 +46,54 @@
 		TakeComponent(detached_pockets)
 
 	U.armor = U.armor.detachArmor(armor)
+	current_uniform = null
```
Comment (Gandalf2k15) [inline code/modules/clothing/under/accessories.dm:49]: SKYRAT EDIT HERE.

File: code/modules/clothing/under/accessories.dm [hunk 6]
```diff
@@ -45,19 +46,54 @@
...
 
 	if(isliving(user))
 		on_uniform_dropped(U, user)
 
 	if(minimize_when_attached)
 		transform *= 2
-		pixel_x -= 8
-		pixel_y += 8
+		pixel_x = 0
```
Comment (Gandalf2k15) [inline code/modules/clothing/under/accessories.dm:56]: SKYRAT EDIT ALL OF THIS

File: code/__DEFINES/combat/block.dm [hunk 7]
```diff
@@ -73,8 +73,9 @@
 // THESE MUST NEVER BE 0! Block code uses ! instead of isnull for the speed boost.
 #define BLOCK_PRIORITY_ACTIVE_BLOCK				200
 #define BLOCK_PRIORITY_HELD_ITEM				100
-#define BLOCK_PRIORITY_CLOTHING					50
 #define BLOCK_PRIORITY_WEAR_SUIT				75
+#define BLOCK_PRIORITY_CLOTHING					50
 #define BLOCK_PRIORITY_UNIFORM					25
+#define BLOCK_PRIORITY_UNDERWEAR				20
```
[the remark this answers is not in the document]
Comment (BobJoga) [reply code/__DEFINES/combat/block.dm:79]: fuck man i kinda forgot to mark the edits lmao

File: code/datums/components/tackle.dm [hunk 8]
```diff
@@ -253,18 +253,23 @@
...
+				defense_mod += 1
+			//
 		if(suit_slot && (istype(suit_slot,/obj/item/clothing/suit/space/hardsuit)))
 			defense_mod += 1
 		if(T.is_shove_knockdown_blocked()) // riot armor and such
 			defense_mod += 5
 		if(T.is_holding_item_of_type(/obj/item/shield))
 			defense_mod += 2
-
+		/* No. No mechanical advantages for liggers.
```
[the remark this answers is not in the document]
Comment (BobJoga) [reply code/datums/components/tackle.dm:266]: Might do this as a separate PR.

File: code/game/objects/structures/dresser.dm (hunk 1 above)
Comment (BobJoga) [reply code/game/objects/structures/dresser.dm:35]: This was all commented out. I'm gonna remove the confusing comment and changes.

File: code/modules/client/preferences.dm [hunk 9]
```diff
@@ -130,6 +130,7 @@ GLOBAL_LIST_INIT(food, list( // Skyrat addition
 
 	var/list/foodlikes = list()
 	var/list/fooddislikes = list()
+	var/list/color_gear = list()
```
[the remark this answers is not in the document]
Comment (BobJoga) [reply code/modules/client/preferences.dm:133]: this is in the skyrat edit section

File: code/modules/client/preferences_savefile.dm [hunk 10]
```diff
@@ -533,6 +533,7 @@ SAVEFILE UPDATING/VERSIONING - 'Simplified', or rather, more coder-friendly ~Car
 	S["feature_horns_color"]			>> features["horns_color"]
 	S["feature_wings_color"]			>> features["wings_color"]
 	//SKYRAT CHANGES
+	S["color_gear"]			>> color_gear
```
[the remark this answers is not in the document]
Comment (BobJoga) [reply code/modules/client/preferences_savefile.dm:536]: already in a skyrat edit section

**Comment by YakumoChen** [comment]
are we going to get extra points to spend on basic decency or is the underwear going to be free?

**Comment by BobJoga** [comment]
> are we going to get extra points to spend on basic decency or is the underwear going to be free?

neither, you have to spend loadout points for underwear

**Comment by DarkWolfInsanity** [comment]
> 
> 
> > are we going to get extra points to spend on basic decency or is the underwear going to be free?
> 
> neither, you have to spend loadout points for underwear

Ver funne much laff. No seriously, underwear is free in character customization, why would I waste more loadout points i could be using to buy me six pens and three plushies for underwear?

**Comment by BobJoga** [comment]
> > > are we going to get extra points to spend on basic decency or is the underwear going to be free?
> > 
> > 
> > neither, you have to spend loadout points for underwear
> 
> Ver funne much laff. No seriously, underwear is free in character customization, why would I waste more loadout points i could be using to buy me six pens and three plushies for underwear?

yeah then don't i guess

**Comment by Owai-Seek** [comment]
You can't preview any of these clothes, and it means you have to strip twice as many things off of a person in medical.

**Comment by BobJoga** [comment]
> You can't preview any of these clothes now, and it means you have to strip twice as many things off of a person in medical and security.
> 
> Supposedly, socks don't work with digigrade legs.
> 
> Underwear show up on people when you examine them as items they have equipped, even though it's under their clothes, further cluttering things when you examine people.
> 
> People have also mentioned accidental prefbreaking because you can now remove peoples underwear, which -you shouldn't be able to do.- Someone is going to use it as an antag gimmick and a whole bunch of nonsense drama is going to get started over being able to nick peoples undergarments through their jumpsuits.

Removing clothing isn't prefbreak.

**Comment by DarkWolfInsanity** [comment]
> 
>
> Someone is going to use it as an antag gimmick and a whole bunch of nonsense drama is going to get started over being able to nick peoples undergarments through their jumpsuits.

Ok, if someone can manage to do that, they have the most godly pick pocketing skill far over Skyrim and Fallout. And yeah. I also agree it's not pref-break cause it's not erp or non-con in the slightest. It's like ripping someone's pants down to their feet in the middle of the hallway. Comedy for you, humiliation for them.

**Comment by Weblure** [comment]
> Ok, if someone can manage to do that, they have the most godly pick pocketing skill far over Skyrim and Fallout.
Thieves' gloves still exist, right? Either way, it's no harder than pickpocketing literally anything else off of someone. In fact, it'd be easier, because they're much less likely to notice it missing. At least with other items, you'll notice something missing from your inventory or from your sprite.

> And yeah. I also agree it's not pref-break cause it's not erp or non-con in the slightest. It's like ripping someone's pants down to their feet in the middle of the hallway. Comedy for you, humiliation for them.

I'm pretty sure that's considered sexual assault in at least a few countries. Either way, that's just going to get annoying fast, and I'm unsure how hard it is for mods to track anyone that does it.

Aside from that, do these extra layers of clothing interfere with medical procedures? This would be very bad because A.) that wastes time and makes an even bigger mess of clothing and B.) a lot of people have underwear to protect their characters from being stipped completely naked, so forcing medical to break prefs to treat their patients is not ideal.

I'm also not a fan of adding more clutter to the UI, and I still don't see the point in having two headset slots.

**Comment by BobJoga** [comment]
> > Removing clothing isn't prefbreak.
> 
> It didn't used to be prefbreak, because people could choose to have underwear on under their clothes which, until this PR, couldn't be removed by other people. Now that underwear can be forcibly removed, it can and should be considered pref-breaking. Don't get me wrong, being able to remove someone else's underwear is a neat feature, but they should have the option to opt-out.
> 
> > Ok, if someone can manage to do that, they have the most godly pick pocketing skill far over Skyrim and Fallout.
> 
> Thieves' gloves still exist, right? Either way, it's no harder than pickpocketing literally anything else off of someone. In fact, it'd be easier, because they're much less likely to notice it missing. At least with other items, you'll notice something missing from your inventory or from your sprite.
> 
> > And yeah. I also agree it's not pref-break cause it's not erp or non-con in the slightest. It's like ripping someone's pants down to their feet in the middle of the hallway. Comedy for you, humiliation for them.
> 
> I'm pretty sure that's considered sexual assault in at least a few countries. Forcing full nudity on someone is borderline forced-ERP and we probably shouldn't toe that line, regardless of your personal opinions on what constitutes as ERP/Non-con. Either way, that's just going to get annoying fast, and I'm unsure how hard it is for mods to track anyone that does it.
> 
> Aside from that, do these extra layers of clothing interfere with medical procedures? This would be very bad because A.) that wastes time and makes an even bigger mess of clothing and B.) a lot of people have underwear to protect their characters from being stipped completely naked, so forcing medical to break prefs to treat their patients is not ideal.
> 
> I'm also not a fan of adding more clutter to the UI, and I still don't see the point in having two headset slots.

Removing clothing isn't prefbreak. You can get free underwear in the loadout. This doesn't clutter the UI since you have a button to hide extra slots. I'm not gonna argue here.

**Comment by Weblure** [comment]
> Removing clothing isn't prefbreak. 

Removing clothing wasn't normally pref break because you had hardcoded, unremovable underwear beneath the clothes. However, since we can now forcibly remove this underwear, that's quite a bit different than removing clothing normally, don't you think?

> You can get free underwear in the loadout. 

That's neat, but that has nothing to do with anything I said.

> This doesn't clutter the UI since you have a button to hide extra slots.

That's fair, now that I've seen it, it's not that bad-looking.

> I'm not gonna argue here.

I thought the Github PR comments section was the prefered place to discuss PRs? Would you prefer to discuss this elsewhere?

**Comment by CameronWoof** [comment]
Underwear shouldn't be visible in the strip menu until the clothing covering it is removed. Same deal that helmets covering your ear and eye slots have. 

That said I'm not sure why Security would need to pull of people's underwear unless there's some way to hide items in them, which would be its own problem.

**Comment by Weblure** [comment]
> Underwear shouldn't be visible in the strip menu until the clothing covering it is removed. Same deal that helmets covering your ear and eye slots have.
> 
> That said I'm not sure why Security would need to pull off people's underwear unless there's some way to hide items in them, which would be its own problem.

Yeah, after testing this out some more, the only issues I really have is the clutter/being able to see underwear in the strip menu and the lack of consent for stripping people completely nude. However, I've yet to test this for medical; if you're required to remove underwear for surgery/medical treatment like Owai said, then that's still a major problem.

**Comment by CameronWoof** [comment]
Pulling off people's bras and underwear definitely should not be a requirement for surgery. I think adding too much mechanical specificity to these is going to turn what should be a cute update into an annoyance.

**Comment by Weblure** [comment]
> you come out of cloning naked

Not if you have underwear set in your character prefs.

> you are born naked, etc.

Personal opinion on nudity isn't relevant here.

> Just deal with it, non-conning people is already against the rules.

Thus why most things require a consent confirmation already, and why requiring medical to force nudity against consent is a bad thing.

File: tgstation.dme [hunk 11]
```diff
@@ -3790,11 +3790,13 @@
 #include "modular_skyrat\code\modules\client\loadout\_medical.dm"
 #include "modular_skyrat\code\modules\client\loadout\_security.dm"
 #include "modular_skyrat\code\modules\client\loadout\backpack.dm"
+#include "modular_skyrat\code\modules\client\loadout\boxers.dm"
 #include "modular_skyrat\code\modules\client\loadout\hands.dm"
 #include "modular_skyrat\code\modules\client\loadout\head.dm"
 #include "modular_skyrat\code\modules\client\loadout\mask.dm"
 #include "modular_skyrat\code\modules\client\loadout\neck.dm"
-#include "modular_skyrat\code\modules\client\loadout\suit.dm"
```
Comment (NopemanMcHalt) [inline tgstation.dme:3797]: Bob, you accidentally unchecked modular suit loadout code.

Comment (BobJoga) [reply tgstation.dme:3797]: fuck lol

**Pull request merged**